tests/cdu/stack/test_utils.py

```diff
@@ -199,3 +203,83 @@ def test_filled_slots_for_active_flow_only_collects_till_top_most_user_flow_fram
     stack = DialogueStack(frames=[another_user_frame, user_frame])
 
     assert filled_slots_for_active_flow(stack, all_flows) == {"foo", "bar"}
+
+
+def test_end_top_user_flow():
+    user_frame = UserFlowStackFrame(
+        flow_id="my_flow", step_id="collect_bar", frame_id="some-frame-id"
+    )
+    pattern_frame = CollectInformationPatternFlowStackFrame(
+        collect="foo", frame_id="some-other-id"
+    )
+    stack = DialogueStack(frames=[user_frame, pattern_frame])
+
+    end_top_user_flow(stack)
+
+    assert len(stack.frames) == 2
+
+    assert stack.frames[0] == UserFlowStackFrame(
+        flow_id="my_flow", step_id="NEXT:END", frame_id="some-frame-id"
+    )
+    assert stack.frames[1] == CollectInformationPatternFlowStackFrame(
+        collect="foo", frame_id="some-other-id", step_id="NEXT:END"
+    )
+
+
+def test_end_top_user_flow_only_ends_topmost_user_frame():
+    user_frame = UserFlowStackFrame(
+        flow_id="my_flow", step_id="collect_bar", frame_id="some-frame-id"
+    )
+    other_user_frame = UserFlowStackFrame(
+        flow_id="my_other_flow", step_id="collect_bar2", frame_id="some-other-id"
+    )
+    stack = DialogueStack(frames=[other_user_frame, user_frame])
+
+    end_top_user_flow(stack)
+
+    assert len(stack.frames) == 2
+
+    assert stack.frames[0] == UserFlowStackFrame(
+        flow_id="my_other_flow", step_id="collect_bar2", frame_id="some-other-id"
+    )
+    assert stack.frames[1] == UserFlowStackFrame(
+        flow_id="my_flow", step_id="NEXT:END", frame_id="some-frame-id"
+    )
+
+
+def test_end_top_user_flow_handles_interrupt_frames():
+    user_frame = UserFlowStackFrame(
+        flow_id="my_flow",
+        step_id="collect_bar",
+        frame_id="some-frame-id",
+        frame_type=FlowStackFrameType.INTERRUPT,
+    )
+    other_user_frame = UserFlowStackFrame(
+        flow_id="my_other_flow", step_id="collect_bar2", frame_id="some-other-id"
+    )
+    stack = DialogueStack(frames=[other_user_frame, user_frame])
+
+    end_top_user_flow(stack)
+
+    assert len(stack.frames) == 2
+
+    assert stack.frames[0] == UserFlowStackFrame(
+        flow_id="my_other_flow",
+        step_id="collect_bar2",
+        frame_id="some-other-id",
+        frame_type=FlowStackFrameType.REGULAR,
+    )
+
+    assert stack.frames[1] == UserFlowStackFrame(
+        flow_id="my_flow",
+        step_id="NEXT:END",
+        frame_id="some-frame-id",
+        frame_type=FlowStackFrameType.REGULAR,
+    )
```

Comment:
depending on how the discussion above goes, this could be removed

Reply:
:+1: